cagnotte_sale/models/sale_order.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV (http://www.acsone.eu)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models, _
+from odoo.fields import first
+from odoo.exceptions import UserError
+
+
+class SaleOrder(models.Model):
+
+    _inherit = 'sale.order'
+
+    has_cagnotte = fields.Boolean(
+        compute="_compute_has_cagnotte",
+        help="This is used to check if sale order lines contain cagnotte",
+    )
+    order_line = fields.One2many(
+        inverse="_inverse_order_line"
+    )
+
+    @api.multi
+    def _reapply_cagnotte(self):
+        for sale in self.filtered('has_cagnotte'):
+            cagnotte_line = first(
+                self.mapped('order_line').filtered('account_cagnotte_id'))
+            cagnotte = cagnotte_line.account_cagnotte_id
+            sale.unset_cagnotte()
+            sale.apply_cagnotte(cagnotte)
+
+    @api.multi
+    def _inverse_order_line(self):
+        # This is triggered on sale order level but is not sufficent
+        # as if the order line is created by another mean (than e.g. interface)
+        # this is not triggered (as well as write())
+        self._reapply_cagnotte()
+
+    @api.multi
+    @api.depends('order_line.account_cagnotte_id')
+    def _compute_has_cagnotte(self):
+        for sale in self.filtered('order_line.account_cagnotte_id'):
+            sale.has_cagnotte = True
+
+    def _get_cagnotte_usable_amount(self, cagnotte):
+        """
+        This is the usable amount of cagnotte
+        :param cagnotte:
+        :return:
+        """
+        if self.amount_total >= cagnotte.solde_cagnotte:
+            return cagnotte.solde_cagnotte
+        return self.amount_total
+
+    @api.multi
+    def unset_cagnotte(self):
+        self.mapped('order_line').filtered('account_cagnotte_id').unlink()
+
+    @api.multi
+    def apply_cagnotte(self, cagnotte):
+        for sale in self:
+            if cagnotte.partner_id != sale.partner_id:
```

Comment:
This won't work. The partner_id on the cagnotte is not mandatory.  When a customer buys a giftcard for someone else, no partner is set.

Reply:
Ok, I had a doubt on that